crates/evm/evm/src/executors/invariant/error.rs

```diff
@@ -202,70 +224,137 @@ impl InvariantFuzzError {
+                    subset
+                })
+                .collect();
+            (powerset, true)
+        } else {
+            let mut rng = thread_rng();
+            (
+                (0..shrink_limit / 3)
+                    .map(|_| {
+                        // select between 1 and calls.len() - 2 number of indices
+                        let amt: usize = rng.gen_range(1..calls.len() - 1);
+                        // construct a random sequence of indices, up to calls.len() - 1 (sample is
+                        // exclusive range and we dont include the last tx
+                        // because its always included), and amt number of indices
+                        let mut seq = seq::index::sample(&mut rng, calls.len() - 1, amt).into_vec();
+                        // sort the indices
+                        seq.sort();
+                        // we add the last tx always because thats what ultimately broke the
+                        // invariant
+                        seq.push(calls.len() - 1);
+                        seq
+                    })
+                    .collect(),
+                false,
+            )
+        };
 
-                removed_calls.push(next_removed);
-            } else {
-                // When the process is restarted, `removed_calls` will be empty.
-                removed_calls.push(anchor + 1);
-            }
+        let new_runs = set.len();
+
+        // just try all of them in parallel
+        set.par_iter().for_each(|use_calls| {
+            self.set_fails_successfully(executor.clone(), calls, use_calls, Arc::clone(&shrunk));
+        });
+
+        // at this point we know shrunk is fine to get the inner value for
+        let shrunk = Arc::<RwLock<Vec<usize>>>::try_unwrap(shrunk).unwrap().into_inner();
+
+        if is_powerset {
+            // a powerset is guaranteed to be smallest subset
+            shrunk
+        } else if shrunk.len() < calls.len() && new_runs + runs < self.shrink_run_limit {
+            let new_calls: Vec<_> = calls
+                .iter()
+                .enumerate()
+                .filter_map(|(i, call)| if shrunk.contains(&i) { Some(call.clone()) } else { None })
+                .collect();
+            let new_calls_idxs =
+                self.try_shrinking_recurse(&new_calls, executor, runs + new_runs, 0);
+            // map back to parent idx, we only have to do this because we gave a new set of calls in
+            // the above function
+            new_calls
+                .iter()
+                .enumerate()
+                .filter_map(|(idx, new_call)| {
+                    if !new_calls_idxs.contains(&idx) {
+                        None
+                    } else {
+                        calls.iter().position(|r| r == new_call)
+                    }
+                })
+                .collect()
+        } else if retries <= 3 && new_runs + runs < self.shrink_run_limit {
```

Comment:
Is there any way to make this whole if else chain simpler? It seems like we have to think about a few different edge cases. If not we should at least comment this further so we know what this is all doing because it feels a bit implicit sometimes